core/presence/presence.go

```diff
@@ -0,0 +1,343 @@
+func New(clock Clock) Recorder {
+	return &recorder{clock: clock}
+}
+
+type recorder struct {
+	mu      sync.Mutex
+	enabled bool
+	clock   Clock
+	entries []Value
+}
+
+// Disable clears the entries and marks the recorder as disabled.
+// Note that all agents will have a recorder, but they are only enabled
+// for API server agents.
+func (r *recorder) Disable() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.enabled = false
+	r.entries = nil
+}
+
+// Enable marks the recorder as enabled.
+func (r *recorder) Enable() {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.enabled = true
+
+}
+
+// IsEnabled returns whether or not the recorder is enabled.
+func (r *recorder) IsEnabled() bool {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	return r.enabled
+}
+
+func (r *recorder) find(server string, id uint64) int {
+	for i, e := range r.entries {
+		if e.Server == server && e.ConnectionID == id {
+			return i
+		}
+	}
+	return -1
+}
+
+// Connect adds an entry for the specified agent.
+func (r *recorder) Connect(server, model, agent string, id uint64, controllerAgent bool, userData string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	pos := r.find(server, id)
+	if pos == -1 {
+		r.entries = append(r.entries, Value{
+			Model:           model,
+			Server:          server,
+			Agent:           agent,
+			ControllerAgent: controllerAgent,
+			ConnectionID:    id,
+			Status:          Alive,
+			UserData:        userData,
+			LastSeen:        r.clock.Now(),
+		})
+	} else {
+		// Need to access the value in the array, not a copy of it.
+		r.entries[pos].Status = Alive
+		r.entries[pos].LastSeen = r.clock.Now()
+	}
+}
+
+// Disconnect removes the entry for the specified connection id.
+func (r *recorder) Disconnect(server string, id uint64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	if pos := r.find(server, id); pos >= 0 {
+		if pos == 0 {
+			r.entries = r.entries[1:]
+		} else {
+			r.entries = append(r.entries[0:pos], r.entries[pos+1:]...)
+		}
+	}
+}
+
+// Activity updates the last seen timestamp for the connection specified.
+func (r *recorder) Activity(server string, id uint64) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	if pos := r.find(server, id); pos >= 0 {
+		r.entries[pos].LastSeen = r.clock.Now()
+	}
+}
+
+// ServerDown marks all connections on the specified server as unknown.
+func (r *recorder) ServerDown(server string) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return
+	}
+
+	for i, value := range r.entries {
+		if value.Server == server {
+			r.entries[i].Status = Missing
+		}
+	}
+}
+
+// UpdateServer replaces all known connections for the specified server with
+// the connections specified. If any of the  connections are not for the
+// specified server, and error is returned.
+func (r *recorder) UpdateServer(server string, connections []Value) error {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	if !r.enabled {
+		return errors.New("recorder not enabled")
```

Comment:
why an error here and a no-op elsewhere?

Reply:
Because this is the only method where an error other that "recorder not enabled" would be returned.

I decided that it would make sense to have an error return on every method just for that.